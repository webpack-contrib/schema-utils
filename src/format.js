// TODO eventually move to `Error/format`
export default function(err) {
  // TODO Better messages, see
  // https://github.com/webpack-contrib/schema-utils/issues/18
  const message = {
    typeof: (err) => `options${err.dataPath} should be a {Function}\n`,
    instanceof: (err) => `options${err.dataPath} should be {RegExp}\n`,
  };

  // TODO handle all ajv.keywords
  switch (err.keyword) {
    case 'typeof':
      return message.typeof(err);
    case 'instanceof':
      return message.instanceof(err);
    default:
      return `options${err.dataPath} ${err.message}\n`;
  }
}
