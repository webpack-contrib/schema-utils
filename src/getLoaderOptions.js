/* eslint-disable no-param-reassign */

const querystring = require('querystring');

const parseJson = require('json-parse-better-errors');

const { validate } = require('./validate');

/**
 * Parses the loader option if necessary, and validates it.
 *
 * @param {string | object | null | undefined} options a raw object, a JSON string, or a query string representing a loader option
 * @param {import("./validate").Schema | null | undefined} schema a schema to validate the option against
 * @return {object} the parsed loader option
 */
module.exports.getLoaderOptions = function getLoaderOptions(options, schema) {
  if (typeof options === 'string') {
    if (options.substr(0, 1) === '{' && options.substr(-1) === '}') {
      try {
        options = parseJson(options);
      } catch (e) {
        throw new Error(`Cannot parse string options: ${e.message}`);
      }
    } else {
      options = querystring.parse(options, '&', '=', {
        maxKeys: 0,
      });
    }
  }

  // eslint-disable-next-line no-undefined
  if (options === null || options === undefined) {
    options = {};
  }

  if (schema) {
    let name = 'Loader';
    let baseDataPath = 'options';
    let match;
    // eslint-disable-next-line no-cond-assign
    if (schema.title && (match = /^(.+) (.+)$/.exec(schema.title))) {
      [, name, baseDataPath] = match;
    }
    // @ts-expect-error
    validate(schema, options, {
      name,
      baseDataPath,
    });
  }

  // @ts-expect-error
  return options;
};
