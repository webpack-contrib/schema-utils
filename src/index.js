const { validate, ValidationError } = require('./validate');
const { getLoaderOptions } = require('./getLoaderOptions');

module.exports = { getLoaderOptions, validate, ValidationError };
