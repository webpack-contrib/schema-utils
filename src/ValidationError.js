import format from './format';

class ValidationError extends Error {
  constructor(errors, name) {
    super();

    this.name = 'ValidationError';

    this.message = `${name || ''} Invalid Options\n\n`;

    errors.forEach((err) => {
      this.message += `${format(err)}`;
    });

    this.errors = errors;

    Error.captureStackTrace(this, this.constructor);
  }
}

export default ValidationError;
