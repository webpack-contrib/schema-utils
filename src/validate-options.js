import path from 'path';
import fs from 'fs';
import Ajv from 'ajv';
import ajvKeywords from 'ajv-keywords';
import weblog from 'webpack-log';
import optionsSchema from '../schema/options.json';
import ValidationError from './ValidationError';

const ajv = new Ajv({
  allErrors: true,
  useDefaults: true,
  errorDataPath: 'property',
});
const defaults = {
  exit: false,
  log: false,
  name: '',
  schema: '',
  target: {},
  throw: true,
};
const logName = 'schema-utils';

ajvKeywords(ajv, ['instanceof', 'typeof']);

const validate = (options) => {
  let schema;

  if (typeof options.schema === 'string') {
    schema = fs.readFileSync(path.resolve(options.schema), 'utf8');
    schema = JSON.parse(schema);
  } else {
    schema = Object.assign({}, options.schema);
  }

  if (!ajv.validate(schema, options.target)) {
    let { errors } = ajv;

    errors = errors.sort((a, b) => {
      const aPath = a.dataPath;
      const bPath = b.dataPath;

      if (aPath.startsWith('[')) {
        if (!bPath.startsWith('[')) {
          return -1;
        }

        return aPath < bPath ? -1 : 1;
      }

      if (aPath.startsWith('.') && bPath.startsWith('.')) {
        return aPath < bPath ? -1 : 1;
      }

      return 0;
    });

    const err = new ValidationError(Object.assign(options, { errors }));

    if (options.throw) {
      throw err;
    }

    if (options.log) {
      const log = weblog({ name: options.name || logName, id: logName });
      log.error(`${err.meta.desc}\n\n${err.format()}\n`);
    }

    if (options.exit) {
      process.exit(1);
    }
  }

  return true;
};

const validator = (opts) => {
  validate({
    throw: true,
    schema: optionsSchema,
    target: opts,
  });

  const options = Object.assign({}, defaults, opts);

  return validate(options);
};

export default validator;
