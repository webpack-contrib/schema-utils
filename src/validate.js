import Ajv from 'ajv';
import ajvKeywords from 'ajv-keywords';

import absolutePathKeyword from './keywords/absolutePath';

import ValidationError from './ValidationError';

const ajv = new Ajv({
  allErrors: true,
  verbose: true,
});

ajvKeywords(ajv, [
  'instanceof',
  'formatMinimum',
  'formatMaximum',
  'patternRequired',
]);

// Custom keywords
absolutePathKeyword(ajv);

function validate(schema, options, configuration = {}) {
  let errors = [];

  if (Array.isArray(options)) {
    errors = options.map((nestedOptions) =>
      validateObject(schema, nestedOptions)
    );

    errors.forEach((list, idx) => {
      const applyPrefix = (error) => {
        // eslint-disable-next-line no-param-reassign
        error.dataPath = `[${idx}]${error.dataPath}`;

        if (error.children) {
          error.children.forEach(applyPrefix);
        }
      };

      list.forEach(applyPrefix);
    });

    errors = errors.reduce((arr, items) => arr.concat(items), []);
  } else {
    errors = validateObject(schema, options);
  }

  if (errors.length > 0) {
    throw new ValidationError(errors, schema, configuration);
  }

  return errors;
}

function validateObject(schema, options) {
  const compiledSchema = ajv.compile(schema);
  const valid = compiledSchema(options);

  return valid ? [] : filterErrors(compiledSchema.errors);
}

function filterErrors(errors) {
  let newErrors = [];

  for (const error of errors) {
    const { dataPath } = error;
    let children = [];

    newErrors = newErrors.filter((oldError) => {
      if (oldError.dataPath.includes(dataPath)) {
        if (oldError.children) {
          children = children.concat(oldError.children.slice(0));
        }

        // eslint-disable-next-line no-undefined, no-param-reassign
        oldError.children = undefined;
        children.push(oldError);

        return false;
      }

      return true;
    });

    if (children.length) {
      error.children = children;
    }

    newErrors.push(error);
  }

  return newErrors;
}

export default validate;
