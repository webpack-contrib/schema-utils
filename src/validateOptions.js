import fs from 'fs';
import path from 'path'; // eslint-disable-line

import Ajv from 'ajv';
import ajvKeywords from 'ajv-keywords';
import ValidationError from './ValidationError';

const ajv = new Ajv({
  useDefaults: true,
  allErrors: true,
  errorDataPath: 'property',
});

ajvKeywords(ajv, ['instanceof', 'typeof']);

const validateOptions = (schema, options, name) => {
  if (typeof schema === 'string') {
    schema = fs.readFileSync(path.resolve(schema), 'utf8'); // eslint-disable-line
    schema = JSON.parse(schema); // eslint-disable-line
  }

  if (!ajv.validate(schema, options)) {
    throw new ValidationError(ajv.errors, name);
  }

  return true;
};

export default validateOptions;
