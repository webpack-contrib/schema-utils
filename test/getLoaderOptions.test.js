import { getLoaderOptions } from '../src';

import schemaTitle from './fixtures/schema-title.json';

describe('getLoaderOptions', () => {
  it('returns given object as-is', () => {
    const options = getLoaderOptions({ name: false });
    expect(options).toEqual({ name: false });
  });

  it('returns and validates given object as-is', () => {
    const options = getLoaderOptions({ name: false }, schemaTitle);
    expect(options).toEqual({ name: false }, schemaTitle);
    expect(() =>
      getLoaderOptions({ namee: false }, schemaTitle)
    ).toThrowErrorMatchingSnapshot();
  });

  it('parses given JSON string', () => {
    const options = getLoaderOptions('{"name":false}');
    expect(options).toEqual({ name: false });
  });

  it('parses and validates given JSON string', () => {
    const options = getLoaderOptions('{"name":false}', schemaTitle);
    expect(options).toEqual({ name: false }, schemaTitle);
    expect(() =>
      getLoaderOptions('{"namee":false}', schemaTitle)
    ).toThrowErrorMatchingSnapshot();
  });

  it('throws an error for invalid JSON', () => {
    expect(() =>
      getLoaderOptions('{"name":faulse}')
    ).toThrowErrorMatchingSnapshot();
  });

  it('parses given query string', () => {
    const options = getLoaderOptions('name=false');
    expect(options).toEqual({ name: 'false' });
  });

  it('parses and validates given query string', () => {
    expect(() =>
      getLoaderOptions('namee=false', schemaTitle)
    ).toThrowErrorMatchingSnapshot();
  });

  it('returns an empty object for null', () => {
    const options = getLoaderOptions(null);
    expect(options).toEqual({});
  });

  it('returns an empty object for undefined', () => {
    // eslint-disable-next-line no-undefined
    const options = getLoaderOptions(undefined);
    expect(options).toEqual({});
  });
});
