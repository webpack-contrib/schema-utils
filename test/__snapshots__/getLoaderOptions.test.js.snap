// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getLoaderOptions parses and validates given JSON string 1`] = `
"Invalid options object. CSS Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'namee'. These properties are valid:
   object { name? }"
`;

exports[`getLoaderOptions parses and validates given query string 1`] = `
"Invalid options object. CSS Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'namee'. These properties are valid:
   object { name? }"
`;

exports[`getLoaderOptions returns and validates given object as-is 1`] = `
"Invalid options object. CSS Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'namee'. These properties are valid:
   object { name? }"
`;

exports[`getLoaderOptions throws an error for invalid JSON 1`] = `"Cannot parse string options: Unexpected token u in JSON at position 10 while parsing near '{\\"name\\":faulse}'"`;
